File: xdl/xdl/core/backend/tf/convert_utils.cc
/* Copyright 2018 Alibaba Group. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "xdl/core/backend/tf/convert_utils.h"
#include "xdl/core/backend/device_singleton.h"
#include "xdl/core/framework/allocator.h"

#include <tensorflow/core/framework/allocator.h>


namespace xdl {

class FakeAllocator:public tensorflow::Allocator {
  public:
    FakeAllocator() {}
    void* AllocateRaw(size_t alignment, size_t num_bytes) override { 
		return buffer_;
	}
			   
	std::string Name() override {
		return "FakeAllocator";
	}
	void DeallocateRaw(void*) override {}
	void SetBuffer(void* buffer) {
		buffer_ = buffer;
		return;
	}
  private:
	void* buffer_ = nullptr;
};

class FakeXDLAllocator:public Allocator {
  public:
    void* Allocate(size_t num_bytes) override {
		return buffer_;
	}
	void Deallocate(void*) override {}
	void SetBuffer(void* buffer) {
		buffer_ = buffer;
		return;
	}
  private:
    void* buffer_ = nullptr;
};


Status XDL2TF::ConvertType(const xdl::DataType s, tensorflow::DataType* d) {
  switch(s) {
  case DataType::kInt8: 
    *d = tensorflow::DT_INT8; 
    return Status::Ok();
  case DataType::kInt16: 
    *d = tensorflow::DT_INT16; 
    return Status::Ok();
  case DataType::kInt32: 
    *d = tensorflow::DT_INT32; 
    return Status::Ok();
  case DataType::kInt64: 
    *d = tensorflow::DT_INT64; 
    return Status::Ok();
  case DataType::kFloat: 
    *d = tensorflow::DT_FLOAT; 
    return Status::Ok();
  case DataType::kDouble: 
    *d = tensorflow::DT_DOUBLE; 
    return Status::Ok();
  default:
    return Status::Internal("unsupported type:" + std::to_string(s));
  }
}

Status XDL2TF::ConvertShape(const xdl::TensorShape& s, tensorflow::TensorShape* d) {
  tensorflow::TensorShape shape;
  for (auto dim: s.Dims()) {
    shape.AddDim(dim);
  }

  *d = shape;
  return Status::Ok();
}

Status XDL2TF::ConvertTensor(const xdl::Tensor& s, tensorflow::Tensor* d) {
  tensorflow::DataType type;
  tensorflow::TensorShape shape;
  XDL_CHECK_STATUS(ConvertType(s.Type(), &type));
  XDL_CHECK_STATUS(ConvertShape(s.Shape(), &shape));
  //avoid memcopy 
  static __thread FakeAllocator fakeAlloc;
  fakeAlloc.SetBuffer(s.Raw<void>());
  *d = tensorflow::Tensor(&fakeAlloc, type, shape);
  return Status::Ok();
}

Status TF2XDL::ConvertType(const tensorflow::DataType s, xdl::DataType* d) {
  switch(s) {
  case tensorflow::DT_INT8: 
    *d = DataType::kInt8; 
    return Status::Ok();
  case tensorflow::DT_INT16: 
    *d = DataType::kInt16; 
    return Status::Ok();
  case tensorflow::DT_INT32: 
    *d = DataType::kInt32; 
    return Status::Ok();
  case tensorflow::DT_INT64: 
    *d = DataType::kInt64; 
    return Status::Ok();
  case tensorflow::DT_FLOAT: 
    *d = DataType::kFloat; 
    return Status::Ok();
  case tensorflow::DT_DOUBLE: 
    *d = DataType::kDouble; 
    return Status::Ok();
  default:
    return Status::Internal("unsupported type:" + std::to_string(s));
  }
}

Status TF2XDL::ConvertShape(const tensorflow::TensorShape& s, xdl::TensorShape* d) {
  std::vector<size_t> dims;
  for (size_t i = 0; i < s.dims(); i++) {
    dims.push_back(s.dim_size(i));
  }

  *d = TensorShape(dims);
  return Status::Ok();
}

Status TF2XDL::ConvertTensor(const tensorflow::Tensor& s, xdl::Tensor* d) {
  DataType type;
  XDL_CHECK_STATUS(ConvertType(s.dtype(), &type));
  TensorShape shape;
  XDL_CHECK_STATUS(ConvertShape(s.shape(), &shape));
  //avoid memcopy
  static __thread FakeXDLAllocator xAllocator;
  xAllocator.SetBuffer((void*)s.tensor_data().data());
  *d = Tensor(&xAllocator, shape, type);
  return Status::Ok();
}

} // namespace xdl

File: xdl/xdl/python/backend/tf/convert_utils.py
# Copyright 2018 Alibaba Group. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import os
import sys

import numpy as np
import tensorflow as tf
from tensorflow.python.ops.init_ops import Constant
from tensorflow.python.ops.init_ops import Ones
from tensorflow.python.ops.init_ops import Zeros
from tensorflow.python.ops.init_ops import RandomNormal
from tensorflow.python.ops.init_ops import RandomUniform
from tensorflow.python.ops.init_ops import TruncatedNormal
from tensorflow.python.ops.init_ops import UniformUnitScaling
from tensorflow.python.ops.init_ops import VarianceScaling

from xdl.python.lib.datatype import DataType as xt
from xdl.python.lib.tensorshape import TensorShape as xts

class XDL2TF(object):
    @staticmethod
    def convert_shape(shape):
        return tf.TensorShape(shape.dims())

    @staticmethod
    def convert_type(dtype):
        if dtype == xt.int16:
            return tf.int16
        if dtype == xt.int32:
            return tf.int32
        elif dtype == xt.int64:
            return tf.int64
        elif dtype == xt.float:
            return tf.float32
        elif dtype == xt.double:
            return tf.float64
        else:
            raise Exception("unknown xdl dtype:", dtype)

class TF2XDL(object):
    @staticmethod
    def convert_shape(shape):
        if isinstance(shape, tf.TensorShape):
            return xts(shape.as_list())
        else:
            return xts(shape)

    @staticmethod
    def convert_type(dtype):
        if dtype == tf.int16:
            return xt.int16
        if dtype == tf.int32:
            return xt.int32
        elif dtype == tf.int64:
            return xt.int64
        elif dtype == tf.float32:
            return xt.float
        elif dtype == tf.float64:
            return xt.double
        else:
            raise Exception("unsupport tf dtype:", dtype)

    @staticmethod
    def convert_initializer(initializer):
        import xdl.python.ops.init_ops as xi
        if initializer == None:
            return xi.Null
        elif type(initializer) == Constant:
            value = getattr(initializer, 'value')
            if isinstance(value, (np.ndarray, np.generic)):
                return xi.Identity(value=value)
            return xi.Constant(value)
        elif type(initializer) == Ones:
            return xi.Ones()
        elif type(initializer) == Zeros:
            return xi.Zeros()
        elif type(initializer) == RandomNormal:
            mean = getattr(initializer, 'mean')
            stddev = getattr(initializer, 'stddev')
            seed = getattr(initializer, 'seed')
            return xi.RandomNormal(mean, stddev, seed)
        elif type(initializer) == RandomUniform:
            minval = getattr(initializer, 'minval')
            maxval = getattr(initializer, 'maxval')
            seed = getattr(initializer, 'seed')
            return xi.RandomUniform(minval, maxval, seed)
        elif type(initializer) == TruncatedNormal:
            mean = getattr(initializer, 'mean')
            stddev = getattr(initializer, 'stddev')
            seed = getattr(initializer, 'seed')
            return xi.TruncatedNormal(mean, stddev, seed)
        elif type(initializer) == UniformUnitScaling:
            factor = getattr(initializer, 'factor')
            seed = getattr(initializer, 'seed')
            return xi.UniformUnitScaling(factor, seed)
        elif type(initializer) == VarianceScaling:
            scale = getattr(initializer, 'scale')
            mode = getattr(initializer, 'mode')
            distribution = getattr(initializer, 'distribution')
            seed = getattr(initializer, 'seed')
            return xi.VarianceScaling(scale, mode, distribution, seed)
        else:
            raise Exception('unsupport tf initializer:' + str(initializer))

